File: csharp/Glacier2/Session/Server/InMemoryPokeStore.cs
// Copyright (c) ZeroC, Inc.

using System.Collections.Immutable;

namespace Server;

/// <summary>An in-memory implementation of <see cref="IPokeStore"/>.</summary>
/// <remarks>This mock implementation is not thread-safe. A real implementation should support concurrent calls.</remarks>
internal class InMemoryPokeStore : IPokeStore
{
    private readonly Dictionary<string, ImmutableList<string>> _store = [];

    /// <inheritdoc />
    public void SaveCollection(string userId, IReadOnlyCollection<string> pokemon)
    {
        if (pokemon.Count == 0)
        {
            _store.Remove(userId);
        }
        else
        {
            _store[userId] = [.. pokemon];
        }
    }

    /// <inheritdoc />
    public IReadOnlyCollection<string>? RetrieveCollection(string userId) =>
        _store.TryGetValue(userId, out ImmutableList<string>? pokemon) ? pokemon : null;
}

File: swift/Glacier2/Session/Sources/Server/InMemoryPokeStore.swift
// Copyright (c) ZeroC, Inc.

import Foundation

/// An in-memory implementation of PokeStore.
/// Note: This mock implementation is not thread-safe. A real implementation should support concurrent calls.
class InMemoryPokeStore: PokeStore {
    private var store: [String: [String]] = [:]

    func saveCollection(userId: String, pokemon: [String]) {
        if pokemon.isEmpty {
            store.removeValue(forKey: userId)
        } else {
            store[userId] = pokemon
        }
    }

    func retrieveCollection(userId: String) -> [String] {
        return store[userId] ?? []
    }
}
